### PR TITLE
Add DynamoDB limiter backend

### PR DESCRIPTION
**Pull request by lamebear**
This adds support to use DynamoDB as a Limiter backend for:
- FixedWindow
- SlidingWindow
- LeakyBucket
- TokenBucket

### CODE REVIEW

File: fixedwindow.go [hunk 1]
```diff
@@ -126,3 +131,76 @@ func (f *FixedWindowRedis) Increment(ctx context.Context, window time.Time, ttl
...
+func (f *FixedWindowDynamoDB) Increment(ctx context.Context, window time.Time, ttl time.Duration) (int64, error) {
+	var resp *dynamodb.UpdateItemOutput
+	var err error
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		resp, err = f.client.UpdateItem(ctx, &dynamodb.UpdateItemInput{
+			Key: map[string]types.AttributeValue{
+				f.tableProps.PartitionKeyName: &types.AttributeValueMemberS{Value: f.partitionKey},
```
Comment (mennanov) [inline fixedwindow.go:171]: Is it possible to use `strconv.FormatInt()` here and in other similar places throughout this PR?
If it is possible, please update the code accordingly

File: fixedwindow.go [hunk 2]
```diff
@@ -126,3 +131,76 @@ func (f *FixedWindowRedis) Increment(ctx context.Context, window time.Time, ttl
...
+	case <-done:
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	}
+
+	if err != nil {
+		return 0, errors.Wrap(err, "dynamodb update item failed")
+	}
+
+	var count float64
```
Comment (mennanov) [inline fixedwindow.go:199]: What's the reason this is `float64` and not `int64`?

File: leakybucket.go [hunk 3]
```diff
@@ -391,3 +395,129 @@ func (t *LeakyBucketRedis) SetState(ctx context.Context, state LeakyBucketState)
...
+		input.ConditionExpression = aws.String(dynamodbBucketRaceConditionExpression)
+		input.ExpressionAttributeValues = map[string]types.AttributeValue{
+			":version": &types.AttributeValueMemberN{Value: strconv.Itoa(int(t.latestVersion))},
+		}
+	}
+
+	return input
+}
+
+func (t *LeakyBucketDynamoDB) getGetItemInput() *dynamodb.GetItemInput {
```
Comment (mennanov) [inline leakybucket.go:501]: Struct LeakyBucketDynamoDB has methods on both value and pointer receivers. Such usage is not recommended by the Go Documentation.

File: tokenbucket.go [hunk 4]
```diff
@@ -465,3 +469,132 @@ func (t *TokenBucketRedis) SetState(ctx context.Context, state TokenBucketState)
 
 	return errors.Wrap(err, "failed to save keys to redis")
 }
+
+// TokenBucketDynamoDB is a DynamoDB implementation of a TokenBucketStateBackend.
+type TokenBucketDynamoDB struct {
+	client        *dynamodb.Client
+	tableProps    DynamoDBTableProperties
+	partitionKey  string
```
Comment (mennanov) [inline tokenbucket.go:477]: How is this `partitionKey` different from the `tableProps.PartitionKeyName`?

File: tokenbucket.go [hunk 5]
```diff
@@ -465,3 +469,132 @@ func (t *TokenBucketRedis) SetState(ctx context.Context, state TokenBucketState)
...
+
+	input := &dynamodb.PutItemInput{
+		TableName: &t.tableProps.TableName,
+		Item:      item,
+	}
+
+	if t.raceCheck && t.latestVersion > 0 {
+		input.ConditionExpression = aws.String(dynamodbBucketRaceConditionExpression)
+		input.ExpressionAttributeValues = map[string]types.AttributeValue{
+			":version": &types.AttributeValueMemberN{Value: strconv.Itoa(int(t.latestVersion))},
```
Comment (mennanov) [inline tokenbucket.go:572]: Redundant type conversion: 

```suggestion
			":version": &types.AttributeValueMemberN{Value: strconv.Itoa(t.latestVersion)},
```

File: leakybucket.go [hunk 6]
```diff
@@ -391,3 +395,129 @@ func (t *LeakyBucketRedis) SetState(ctx context.Context, state LeakyBucketState)
 
 	return errors.Wrap(err, "failed to save keys to redis")
 }
+
+// LeakyBucketDynamoDB is a DyanamoDB implementation of a LeakyBucketStateBackend.
+type LeakyBucketDynamoDB struct {
+	client        *dynamodb.Client
+	tableProps    DynamoDBTableProperties
+	partitionKey  string
```
Comment (mennanov) [inline leakybucket.go:403]: How is this `paritionKey` different from `tableProps.ParitionKeyName`?

File: dynamodb.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+	//SortKeyUsed indicates if a SortKey is present on the table.
+	SortKeyUsed bool
+	//TTLFieldName is the name of the attribute configured for TTL.
+	TTLFieldName string
+	//TTLUsed indicates if TTL is configured on the table.
+	TTLUsed bool
+}
+
+var tableCache = make(map[string]DynamoDBTableProperties)
+var lock = &sync.Mutex{}
```
Comment (mennanov) [inline dynamodb.go:31]: Using global variables is discouraged. 
I'd suggest removing them and instead make the client code implement the cache if needed.
Same goes for the mutex: let the client code control the concurrent execution behavior.

File: dynamodb.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+	return data, nil
+}
+
+func populateTableTTL(data DynamoDBTableProperties, ttl *types.TimeToLiveDescription) DynamoDBTableProperties {
+	if ttl.TimeToLiveStatus != types.TimeToLiveStatusEnabled {
+		return data
+	}
+
+	data.TTLUsed = true
```
Comment (mennanov) [inline dynamodb.go:113]: Is this field ever used?

**Review by mennanov — CHANGES_REQUESTED** [review]

File: tokenbucket.go (hunk 4 above)
Comment (lamebear) [reply tokenbucket.go:477]: See above comment

File: leakybucket.go (hunk 3 above)
Comment (lamebear) [reply leakybucket.go:501]: All methods are on pointer receivers.

File: dynamodb.go (hunk 7 above)
Comment (lamebear) [reply dynamodb.go]: removed

File: tokenbucket.go (hunk 5 above)
Comment (lamebear) [reply tokenbucket.go]: removed

File: fixedwindow.go (hunk 1 above)
Comment (lamebear) [reply fixedwindow.go:171]: done

**Review by mennanov — CHANGES_REQUESTED** [review]

**Comment by mennanov** [comment]
Is it possible to improve test coverage for `dynamodb.go`?

**Comment by lamebear** [comment]
The `dynamodb.go` file is showing 81.1% coverage locally, 75% prior to this latest commit. Don't know why codecov is only showing 61% prior to this last commit. Any more tests on that file will be just for the sake of coverage percentage and not gaining value.

**Review by mennanov — APPROVED** [review]

**Comment by mennanov** [comment]
Awesome! Thanks for your work!

**Pull request merged**